mypy/checker.py

```diff
@@ -257,21 +258,60 @@ def check_overlapping_overloads(self, defn: OverloadedFuncDef) -> None:
                     self.msg.overloaded_signatures_overlap(i + 1, i + j + 2,
                                                            item.func)
 
-    def is_generator_return_type(self, typ: Type) -> bool:
-        return is_subtype(self.named_generic_type('typing.Generator',
-                                                  [AnyType(), AnyType(), AnyType()]),
-                          typ)
+    # Here's the scoop about generators and coroutines.
+    #
+    # There are two kinds of generators: classic generators (functions
+    # with `yield` or `yield from` in the body) and coroutines
+    # (functions declared with `async def`).  The latter are specified
+    # in PEP 492 and only available in Python >= 3.5.
+    #
+    # Classic generators can be parameterized with three types:
+    # - ty is the yield type (the type of y in `yield y`)
+    # - ts is the type receive by yield (the type of s in `s = yield`)
+    # - tr is the return type (the type of r in `return r`)
+    #
+    # A classic generator must define a return type that's either
+    # `Generator[ty, ts, tr]`, Iterator[ty], or Iterable[ty] (or
+    # object or Any).  If ts/tr are not given, both are Any.
```

Comment:
Actually in the code below ts is Void, tr is Any. But I wonder if it isn't better to make both None -- forcing developers to declare the return type as `Generator[ty, ts, tr]` if they want to either receive or return values. There's a unit test that currently checks that `return value` is allowed in a generator declared as `Iterator` but I think I'll just change that too.